hls-graph/src/Development/IDE/Graph/Internal/Database.hs

```diff
@@ -442,34 +474,62 @@ transitiveDirtyListBottomUp database seeds = do
--- the rights are new affected keys, we need to mark them dirty
-transitiveDirtyListBottomUpDiff :: Foldable t => Database -> t Key -> [Key] -> STM ([Key], [Key], KeySet)
-transitiveDirtyListBottomUpDiff database seeds allOldKeys = do
-  acc <- newTVar []
-  let go1 x = do
-        seen <- State.get
-        if x `memberKeySet` seen
-          then pure ()
-          else do
-            State.put (insertKeySet x seen)
-            when (not (isRootKey x)) $ do
-                mnext <- lift $ getRunTimeRDeps database x
-                traverse_ go1 (maybe mempty toListKeySet mnext)
-            lift $ modifyTVar acc (x :)
-  -- traverse all seeds
-  seen <- snd <$> State.runStateT (do traverse_ go1 seeds) mempty
-  newKeys <- readTVar acc
-  let oldKeys = filter (`notMemberKeySet` seen) allOldKeys
-  return (oldKeys, newKeys, seen)
+-- | A concurrent variant of 'transitiveDirtyListBottomUp' that computes the difference
+-- between two sets of affected keys.
+--
+-- Returns:
+-- * Right keys: newly affected keys that need to be marked dirty
+-- * Left keys: previously affected keys that are no longer affected (can be marked clean)
+--
+-- The function traverses the reverse-dependency graph concurrently, processing independent
+-- branches in parallel while maintaining bottom-up ordering (dependencies before dependents).
+-- This improves performance on large dependency graphs by utilizing multiple cores.
+--
+-- Thread-safety is ensured by:
+-- * Using TVar for shared state (visited set and accumulator)
+-- * Atomic check-and-mark for the visited set
+-- * mapConcurrently for parallel traversal of independent branches
+transitiveDirtyListBottomUpDiff :: Database -> [Key] -> [Key] -> IO [Either Key Key]
+transitiveDirtyListBottomUpDiff database seeds lastSeeds = do
+  -- Use TVars for thread-safe concurrent access
+  accTVar <- newTVarIO []
+  seenTVar <- newTVarIO mempty
+  
+  let -- Process a key and its dependencies concurrently
+      go :: (Key -> Either Key Key) -> Key -> IO ()
+      go wrapper x = do
+        alreadySeen <- atomically $ do
+          seen <- readTVar seenTVar
+          if x `memberKeySet` seen
+            then pure True
+            else do
+              writeTVar seenTVar (insertKeySet x seen)
+              pure False
+        
+        unless alreadySeen $ do
+          -- Fetch dependencies
+          mnext <- atomically $ getRunTimeRDeps database x
+          let deps = maybe [] toListKeySet mnext
+          
+          -- Process dependencies concurrently
+          unless (null deps) $ do
+            void $ mapConcurrently (go wrapper) deps
+          
+          -- Add this key to accumulator after all dependencies are processed
+          atomically $ modifyTVar' accTVar (wrapper x :)
+  
+  -- Process new seeds (Right) and old seeds (Left) concurrently
+  void $ mapConcurrently (go Right) seeds
+  void $ mapConcurrently (go Left) lastSeeds
+  
+  readTVarIO accTVar
```

Comment:
This is actually slower since we are using too many threads. We might do a bfs, using limited number of threads to speed up the computation